### PR TITLE
Improve the standalone kube-aggregator

### PR DESCRIPTION
**Pull request by bjrara**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

According to the design-proposal of the kube-aggregator, it supports running in two modes: one is test mode/single-user mode, and the other is gateway mode. However, the first mode doesn't work in today's code base.

This PR fixes issues of deploying a standalone kube-aggregator.

Prerequiste PR:
https://github.com/kubernetes/kubernetes/pull/91537

**Description:**
When we deploy a standalone kube-aggregator, we lose the aggregation layer setup which originally managed by the kube-apiserver:
* OPENAPI Configuration;
* Customization of the ServerRunOption;

Besides, the kube-aggregator fails to handle the situation where it is standalone deployed and there's no master apiserver is in-process running with it:
* APIService health check failure to the legacy group v1: the legacy group v1 doesn’t conform to the health check convention of APIServices.

**Which issue(s) this PR fixes**:

Fixes #

* Add default OPENAPI configuration to the kube-aggregator in the standalone mode to run the openAPI aggregator which installs an openAPI handler to the server container.
* Add ServerRunOptions to the `AggregatorOptions` used in the standalone kube-aggregator to enable customization of preflight handlers, e.g. max-requests-inflight.

* Fix the health check failure regarding the legacy group v1.

**Special notes for your reviewer**:

Changes in this PR will not affect the kube-aggregator in the gateway mode.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: build/root/Makefile.generated_files [hunk 1]
```diff
@@ -501,23 +501,26 @@ API_RULE_CHECK_FAILURE_MESSAGE = "ERROR: \n\t $(1) API rule check failed. Report
...
 KUBE_KNOWN_VIOLATION_FILENAME := api/api-rules/violation_exceptions.list
+AGGREGATOR_KNOWN_VIOLATION_FILENAME := api/api-rules/aggregator_violation_exceptions.list
 APIEXTENSIONS_KNOWN_VIOLATION_FILENAME := api/api-rules/apiextensions_violation_exceptions.list
 CODEGEN_KNOWN_VIOLATION_FILENAME := api/api-rules/codegen_violation_exceptions.list
 SAMPLEAPISERVER_KNOWN_VIOLATION_FILENAME := api/api-rules/sample_apiserver_violation_exceptions.list
 
 APIMACHINERY_DEFAULT_TAG_FILES := vendor/k8s.io/apimachinery/pkg/apis/meta/v1/% vendor/k8s.io/apimachinery/pkg/runtime/% vendor/k8s.io/apimachinery/pkg/version/%
 
 KUBE_OPENAPI_TAG_FILES :=            $(filter-out vendor/k8s.io/code-generator/% vendor/k8s.io/sample-apiserver/%, $(ALL_K8S_TAG_FILES))
+AGGREGATOR_OPENAPI_TAG_FILES :=      $(filter $(APIMACHINERY_DEFAULT_TAG_FILES) vendor/k8s.io/kube-aggregator/%, $(ALL_K8S_TAG_FILES))
```
Comment (yue9944882) [inline build/root/Makefile.generated_files:512]: openapi schemas for aggregator is for now generated/embeded from the kube-apiserver https://github.com/kubernetes/kubernetes/pull/50864#discussion_r133905283. moving aggregator schemas to a separate package is probably a positive change.

**Review by yue9944882 — COMMENTED** [review]
the proposed change from this pull looks fractional to me and i didnt see the intention from the diff

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 2]
```diff
@@ -156,5 +167,10 @@ func (o AggregatorOptions) RunAggregator(stopCh <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
-	return server.GenericAPIServer.PrepareRun().Run(stopCh)
+
+	prepared, err := server.PrepareRun()
```
[the remark this answers is not in the document]
Comment (bjrara) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go:171]: The openapiConfig is assigned to the APIAggregator (`server` in the code), prepareRun the genericAPIServer won't invoke openapi setup, but prepareRun the server will do.

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ import (
 
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
+	openapinamer "k8s.io/apiserver/pkg/endpoints/openapi"
+	"k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/generated/openapi"
```
[the remark this answers is not in the document]
Comment (bjrara) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: fixed. thx.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 4]
```diff
@@ -285,7 +285,13 @@ func (c *AvailableConditionController) sync(key string) error {
 					results <- err
 					return
 				}
-				discoveryURL.Path = "/apis/" + apiService.Spec.Group + "/" + apiService.Spec.Version
+				// skip legacyAPIService health check when it is delegated to the core apiserver
+				if apiService.Name == "v1." {
```
[the remark this answers is not in the document]
Comment (bjrara) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:289]: The aggregator skips health check of apiservices marked local. This line will only be executed when the service reference is not nil, which means delegating to a service.

If we configure APIService v1. with a service reference, this health check will always fail (as the legacy group doesn't apply to the health check rule here). That's why I add the skip check.

File: build/root/Makefile.generated_files (hunk 1 above)
Comment (bjrara) [reply build/root/Makefile.generated_files:512]: Separated OPENAPI change to a new PR https://github.com/kubernetes/kubernetes/pull/91537

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 5]
```diff
@@ -77,6 +80,7 @@ func NewCommandStartAggregator(defaults *AggregatorOptions, stopCh <-chan struct
 
 // AddFlags is necessary because hyperkube doesn't work using cobra, so we have to have different registration and execution paths
 func (o *AggregatorOptions) AddFlags(fs *pflag.FlagSet) {
+	o.ServerRunOptions.AddUniversalFlags(fs)
```
Comment (lavalamp) [inline staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go:83]: Won't this double-register the flags in the kube-apiserver binary?

Comment (lavalamp) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go:83]: Or, wait, maybe this is correct, I need to double check which layer kube-apiserver imports

Comment (bjrara) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go:83]: It won't. The AggregatorOptions is used when the kube-aggregator starts standalone.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lavalamp) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:289]: Is this the correct fix? I guess I would have expected it to proceed, but with a different path?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 6]
```diff
@@ -285,7 +285,13 @@ func (c *AvailableConditionController) sync(key string) error {
 					results <- err
 					return
 				}
-				discoveryURL.Path = "/apis/" + apiService.Spec.Group + "/" + apiService.Spec.Version
+				// skip legacyAPIService health check when it is delegated to the core apiserver
```
Comment (lavalamp) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:288]: This comment is confusing-- we already know the service isn't local due to line 278? So how can this be delegated to the core apiserver?

Comment (bjrara) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: Changed the comment to `skip legacyAPIService health check when it is delegated to a service`. Hope it better describe the intention.

**Comment by lavalamp** [comment]
Let's get #91537 merged, and then I think we can rebase this one and get it merged.

**Comment by lavalamp** [comment]
Thanks for sending these PRs in digestible chunks rather than a single giant change!

**Pull request merged**